### PR TITLE
rgw: fix bug deleting a non-existent object also generates a delete marker

### PR DESCRIPTION
**Pull request by kxkxk**
fix:https://tracker.ceph.com/issues/56992
Signed-off-by: Yuxuan Yang [xuanyu.yang@sincerecloud.com](mailto:xuanyu.yang@sincerecloud.com)
I don't think rgw should create a delete marker for a non-existent object

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -5013,6 +5013,13 @@ void RGWDeleteObj::execute(optional_yield y)
         etag = astate->attrset[RGW_ATTR_ETAG].to_str();
       }
 
+      // if the object does not exist, it is not allowed to create a delete_marker
+      if(s->bucket->get_info().versioning_enabled()&&!s->object->have_instance()){
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc:5017]: you need white-space around operators && and ||, and later, after the arguments to a function.

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -5013,6 +5013,13 @@ void RGWDeleteObj::execute(optional_yield y)
...
+        rgw_bucket_olh_entry olh;
+        op_ret = store->getRados()->bi_get_olh(this,s->bucket->get_info(),s->object->get_obj(),&olh);
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc:5019]: I'm not sure what this check actually does.  could someone please explain what calling get_olh() on this object (without intent to use the result) accomplishes?

**Review by ivancich — CHANGES_REQUESTED** [review]
I'm also wondering if this is where we should code this. I have a merged PR where we would not add a delete marker after an existing delete marker (see: https://github.com/ceph/ceph/pull/41897/files). There is existing code is src/cls/rgw/cls_rgw.cc in function rgw_bucket_link_olh, which loads in the OLH. In this area (https://github.com/ceph/ceph/blob/ee321e85703580c8e5c4575342ff753afa1fe20d/src/cls/rgw/cls_rgw.cc#L1763) if olh_found is false, we have a case where we're trying to add a delete marker when there is no object.

And that would put this adjacent to the stacked delete marker checks.

What do you think?

File: src/rgw/rgw_op.cc (hunk 1 above)
Comment (kxkxk) [reply src/rgw/rgw_op.cc:5017]: oops, thanks for pointing out. I'll update it after I fix the error for make check

File: src/rgw/rgw_op.cc (hunk 2 above)
Comment (kxkxk) [reply src/rgw/rgw_op.cc:5019]: This call is used to determine whether the object exists. If the object does not exist, the object's olh should not exist, then ret < 0

**Review by kxkxk — COMMENTED** [review]
I modified this part of the code so that the function is called correctly. My intention is that the bucket does not create delete markers for objects that are not created. This will be achieved by determining whether OLH exists or not.

File: src/cls/rgw/cls_rgw.cc [hunk 3]
```diff
@@ -1759,6 +1759,11 @@ static int rgw_bucket_link_olh(cls_method_context_t hctx, bufferlist *in, buffer
     if (ret < 0) {
       return ret;
     }
+
+    // if the OLH not exist, error out
```
Comment (ivancich) [inline src/cls/rgw/cls_rgw.cc:1763]: Maybe add "does" as in "if the OLH does not exist, error out".

File: src/cls/rgw/cls_rgw.cc [hunk 4]
```diff
@@ -1780,7 +1785,8 @@ static int rgw_bucket_link_olh(cls_method_context_t hctx, bufferlist *in, buffer
       unmod.tv_nsec = 0;
     }
     if (mtime >= unmod) {
-      return 0; /* no need tof set error, we just return 0 and avoid
+      return 0;
+     /* no need tof set error, we just return 0 and avoid
```
Comment (ivancich) [inline src/cls/rgw/cls_rgw.cc:1789]: I realize you didn't create this misspelling, but since you're editing here anyway maybe change "tof" to "to".

**Review by ivancich — APPROVED** [review]
Thank you!

File: src/cls/rgw/cls_rgw.cc (hunk 3 above)
Comment (kxkxk) [reply src/cls/rgw/cls_rgw.cc]: Ok, thanks for pointing out

**Review by kxkxk — COMMENTED** [review]
Correct comment errors

**Comment by ivancich** [comment]
I just realized this is 10 commits. Are you familiar with how to squash all this down to a single commit? If you'd like some guidance, just let me know. I've added a DNM (do not merge) tag so prevent merge until the squash. But QA can proceed.

Also when you're down to one commit, it needs to be signed, as the CI is currently complaining about an unsigned commit.

Thanks!

**Comment by kxkxk** [comment]
> I just realized this is 10 commits. Are you familiar with how to squash all this down to a single commit? If you'd like some guidance, just let me know. I've added a DNM (do not merge) tag so prevent merge until the squash. But QA can proceed.
> 
> Also when you're down to one commit, it needs to be signed, as the CI is currently complaining about an unsigned commit.
> 
> Thanks!

Some merge operations are included in these commits and I don't konw how to squash them, what should I do？
Thanks for your help!

**Comment by ivancich** [comment]
> > I just realized this is 10 commits. Are you familiar with how to squash all this down to a single commit? If you'd like some guidance, just let me know. I've added a DNM (do not merge) tag so prevent merge until the squash. But QA can proceed.
> > Also when you're down to one commit, it needs to be signed, as the CI is currently complaining about an unsigned commit.
> > Thanks!
> 
> Got it. Thanks for your advice!

Great, thank you. I don't know which method you used, but my tool for this is `git rebase -i HEAD~_N_`, where N is the number of commits. There are special occasions where I use `git rebase -ir ...`.

**Comment by ivancich** [comment]
@kxkxk, I updated the tracker for backports to quincy and pacific. Please edit if you think it should be different. Thanks!

**Comment by kxkxk** [comment]
> @kxkxk, I updated the tracker for backports to quincy and pacific. Please edit if you think it should be different. Thanks!

It is available on both versions.

**Review by kxkxk — COMMENTED** [review]
Thanks! I think it's ready.

**Comment by cbodley** [comment]
this failed qa on the s3test case [test_versioning_multi_object_delete_with_marker_create](https://github.com/ceph/s3-tests/blob/master/s3tests_boto3/functional/test_s3.py#L7522-L7538):
```
=================================== FAILURES ===================================
____________ test_versioning_multi_object_delete_with_marker_create ____________

    @pytest.mark.fails_on_dbstore
    def test_versioning_multi_object_delete_with_marker_create():
        bucket_name = get_new_bucket()
        client = get_client()

        check_configure_versioning_retry(bucket_name, "Enabled", "Enabled")

        key = 'key'

        response = client.delete_object(Bucket=bucket_name, Key=key)
>       delete_marker_version_id = response['VersionId']
E       KeyError: 'VersionId'

s3tests_boto3/functional/test_s3.py:7531: KeyError
```
i tested amazon's behavior with the following boto script, and see that they allow this creation of delete markers:
```python
client.delete_object(Bucket='test-nonexistent-delete-marker', Key='nonexistent-key')
response = client.list_object_versions(Bucket='test-nonexistent-delete-marker')
print(response['DeleteMarkers'])
```
> [{'Owner': {'ID': '...'}, 'Key': 'nonexistent-key', 'VersionId': 'H8Czx7yv25IqznMOQbbU4bU.iHKfSqSn', 'IsLatest': True, 'LastModified': datetime.datetime(2023, 3, 13, 14, 55, 48, tzinfo=tzutc())}]

**Comment by kxkxk** [comment]
> this failed qa on the s3test case [test_versioning_multi_object_delete_with_marker_create](https://github.com/ceph/s3-tests/blob/master/s3tests_boto3/functional/test_s3.py#L7522-L7538):
> 
> ```
> =================================== FAILURES ===================================
> ____________ test_versioning_multi_object_delete_with_marker_create ____________
> 
>     @pytest.mark.fails_on_dbstore
>     def test_versioning_multi_object_delete_with_marker_create():
>         bucket_name = get_new_bucket()
>         client = get_client()
> 
>         check_configure_versioning_retry(bucket_name, "Enabled", "Enabled")
> 
>         key = 'key'
> 
>         response = client.delete_object(Bucket=bucket_name, Key=key)
> >       delete_marker_version_id = response['VersionId']
> E       KeyError: 'VersionId'
> 
> s3tests_boto3/functional/test_s3.py:7531: KeyError
> ```
> 
> i tested amazon's behavior with the following boto script, and see that they allow this creation of delete markers:
> 
> ```python
> client.delete_object(Bucket='test-nonexistent-delete-marker', Key='nonexistent-key')
> response = client.list_object_versions(Bucket='test-nonexistent-delete-marker')
> print(response['DeleteMarkers'])
> ```
> 
> > [{'Owner': {'ID': '...'}, 'Key': 'nonexistent-key', 'VersionId': 'H8Czx7yv25IqznMOQbbU4bU.iHKfSqSn', 'IsLatest': True, 'LastModified': datetime.datetime(2023, 3, 13, 14, 55, 48, tzinfo=tzutc())}]

It should be that I misunderstood the AWS API documentation related to this. I don't understand the point of creating delete markers for non-existent objects. Maybe we should keep pace with Amazon?

**Comment by NarrativeBias** [comment]
@kxkxk, mind checking something semi-related to this issue for me?
Those delete markers for non-existent objects are generally successfully cleaned up by the ExpiredObjectDeleteMarker lifecycle however when I create that marker by sending two delete requests and process that lifecycle - there are leftover objects for those markers in the data pool.

**Pull request closed without merging**